lattices/src/map_union_with_tombstones.rs

```diff
@@ -0,0 +1,443 @@
+}
+
+impl<MapSelf, MapOther, K, ValSelf, ValOther, TombstoneSetSelf, TombstoneSetOther>
+    PartialOrd<MapUnionWithTombstones<MapOther, TombstoneSetOther>>
+    for MapUnionWithTombstones<MapSelf, TombstoneSetSelf>
+where
+    MapSelf: Map<K, ValSelf, Key = K, Item = ValSelf> + MapIter + SimpleKeyedRef,
+    MapOther: Map<K, ValOther, Key = K, Item = ValOther> + MapIter + SimpleKeyedRef,
+    ValSelf: PartialOrd<ValOther> + IsBot,
+    ValOther: IsBot,
+    TombstoneSetSelf: Len + Iter<Item = K> + for<'a> Get<&'a K>,
+    TombstoneSetOther: Len + Iter<Item = K> + for<'a> Get<&'a K>,
+{
+    fn partial_cmp(
+        &self,
+        other: &MapUnionWithTombstones<MapOther, TombstoneSetOther>,
+    ) -> Option<Ordering> {
+        let self_tombstones_greater = self
+            .tombstones
+            .iter()
+            .any(|k| !other.tombstones.contains(&*k));
+
+        let other_tombstones_greater = other
+            .tombstones
+            .iter()
+            .any(|k| !self.tombstones.contains(&*k));
+
+        if self_tombstones_greater && other_tombstones_greater {
+            return None;
+        }
+
+        let mut self_any_greater = false;
+        let mut other_any_greater = false;
+        let self_keys = self
+            .map
+            .iter()
+            .filter(|(k, v)| {
+                !v.is_bot() && !self.tombstones.contains(k) && !other.tombstones.contains(k)
+            })
+            .map(|(k, _v)| <MapSelf as SimpleKeyedRef>::into_ref(k));
+        let other_keys = other
+            .map
+            .iter()
+            .filter(|(k, v)| {
+                !v.is_bot() && !self.tombstones.contains(k) && !other.tombstones.contains(k)
+            })
+            .map(|(k, _v)| <MapOther as SimpleKeyedRef>::into_ref(k));
+
+        for k in self_keys.chain(other_keys) {
+            match (self.map.get(k), other.map.get(k)) {
+                (Some(self_value), Some(other_value)) => {
+                    match self_value.partial_cmp(&*other_value) {
+                        None => {
+                            return None;
+                        }
+                        Some(Less) => {
+                            other_any_greater = true;
+                        }
+                        Some(Greater) => {
+                            self_any_greater = true;
+                        }
+                        Some(Equal) => {}
+                    }
+                }
+                (Some(_), None) => {
+                    self_any_greater = true;
+                }
+                (None, Some(_)) => {
+                    other_any_greater = true;
+                }
+                (None, None) => unreachable!(),
+            }
+            if self_any_greater && other_any_greater {
+                return None;
+            }
+        }
+        match (
+            self_any_greater,
+            other_any_greater,
+            self_tombstones_greater,
+            other_tombstones_greater,
```

Comment:
I think `self_any_greater` and `self_tombstones_greater` could be combined into a single variable? And same for `other_`